### PR TITLE
GH-43677: [C++][FlightRPC] Move the FlightTestServer to its own .cc and .h files

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

One way of learning about a codebase is reading the tests. As it is now, it's hard to see the minimal `FlightServerBase` sub-class in `flight/test_util.cc`, so I moved it to its own file.

### What changes are included in this PR?

 - Renaming `FlightTestServer` to `TestFlightServer`
 - Moving the class to `test_flight_server.{h,cc}`
 - Bonus: Moving the server and client auth handlers to `test_auth_handlers.{h,cc}`

### Are these changes tested?

By existing tests.

### Are there any user-facing changes?

`ExampleTestServer` is removed from the testing library in favor of `FlightTestServer::Make`.
* GitHub Issue: #43677

### CODE REVIEW

**Comment by felipecrv** [comment]
Pushed a commit moving the auth handlers as well.

**Review by lidavidm — APPROVED** [review]
I'll just note that these are very much not meant as pedagogical examples

File: cpp/src/arrow/flight/test_flight_server.cc [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+  } else {
+    return Status::NotImplemented("no stream implemented for ticket: " + ticket.ticket);
+  }
+}
+
+}  // namespace
+
+std::unique_ptr<FlightServerBase> TestFlightServer::Make() {
+  return std::make_unique<TestFlightServer>();
+}
```
Comment (kou) [inline cpp/src/arrow/flight/test_flight_server.cc:83]: Do we need this factory method?
How about using `std::make_unique<TestFlightServer>()` directly in callers?

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/flight/test_flight_server.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/flight/test_flight_server.cc:83]: It's used quite a lot and is handy to not have to write the longer make_unique invocation. This is replacing the old free function called `ExampleTestServer`.

**Pull request merged**